### PR TITLE
[TML] Introduce belowground competition concept 'symmetric ZOI'

### PR DESCRIPTION
**Pull request by mcwimm**
Below-ground competition concept based on BETTINA ODD 2017:
symmetric zone of influence (no salinity)

### CODE REVIEW

File: TreeModelLib/BelowgroundCompetition/__init__.py [hunk 1]
```diff
@@ -18,3 +18,4 @@
 from .NetworkOGSLargeScale3D import NetworkOGSLargeScale3D
 from .OGSLargeScale3DExternal import OGSLargeScale3DExternal
 from .NetworkOGSLargeScale3DExternal import NetworkOGSLargeScale3DExternal
+from. SymmetricZOI import SymmetricZOI
```
Comment (jvollhueter) [inline TreeModelLib/BelowgroundCompetition/__init__.py:21]:
```suggestion
from .SymmetricZOI import SymmetricZOI
```

**Review by jvollhueter — CHANGES_REQUESTED** [review]
Looks nice! The only thing I noticed is a typo.

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    ## SymmetricZOI case for below ground competition concept. Symmetric
+    #  Zone Of Influence with trees occupying the same node of the grid share
+    #  the below-ground resource of this node equally (BETTINA geometry of a
+    #  tree assumed). See Peters 2017: ODD protocol of the model BETTINA IBM\n
+    #  @param Tags to define SimpleAsymmetricZOI: see tag documentation
+    #  @date: 2019 - Today
+    def __init__(self, args):
+        case = args.find("type").text
+        print("Initiate belowground competition of type " + case + ".")
+        self.makeGrid(args)
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:23]: Can we move the makeGrid function to TreeModel?

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    #  concept. In the SymmetricZOI concept, trees geometric measures
+    #  are saved in simple lists and the timestepping is updated. \n
+    #  @param t_ini - initial time for next timestep \n
+    #  @param t_end - end time for next timestep
+    def prepareNextTimeStep(self, t_ini, t_end):
+        self.xe = []
+        self.ye = []
+        self.r_root = []
+        self.t_ini = t_ini
+        self.t_end = t_end
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:35]: Move self.xe/ye and times to TreeModel?

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    def calculateBelowgroundResources(self):
+        # Numpy array of shape [res_x, res_y, n_trees]
+        distance = (((self.my_grid[0][:, :, np.newaxis] -
+                      np.array(self.xe)[np.newaxis, np.newaxis, :])**2 +
+                     (self.my_grid[1][:, :, np.newaxis] -
+                      np.array(self.ye)[np.newaxis, np.newaxis, :])**2)**0.5)
+        # Array of shape distance, indicating which cells are occupied by
+        # tree root plates
+        root_radius = np.array(self.r_root)
+        trees_present = root_radius[np.newaxis, np.newaxis, :] > distance
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:71]: Maybe add a comment expalinin the array

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    ## This function returns the BelowgroundResources calculated in the
+    #  subsequent timestep.\n
+    #  @return: np.array with $N_tree$ scalars
+    def calculateBelowgroundResources(self):
+        # Numpy array of shape [res_x, res_y, n_trees]
+        distance = (((self.my_grid[0][:, :, np.newaxis] -
+                      np.array(self.xe)[np.newaxis, np.newaxis, :])**2 +
+                     (self.my_grid[1][:, :, np.newaxis] -
+                      np.array(self.ye)[np.newaxis, np.newaxis, :])**2)**0.5)
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:67]: This array may kill the ram. Can you avoid arrays this big?

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py (hunk 2 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:23]: Good idea!

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py (hunk 3 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:35]: Yes, except for r_root as this variable might not exist in all concepts (e.g. Kiwi), right?

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py (hunk 5 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:67]: Yes. But let's put this in a new issue/PR as this kind of arrays occur in multiple moduls.

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py (hunk 3 above)
Comment (jbathmann) [reply TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:35]: I agree

File: TreeModelLib/BelowgroundCompetition/__init__.py (hunk 1 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/__init__.py]: solved

File: TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py (hunk 3 above)
Comment (mcwimm) [reply TreeModelLib/BelowgroundCompetition/SymmetricZOI/SymmetricZOI.py:35]: see issue #147

**Comment by jbathmann** [comment]
Looks good.
If @jvollhueter agrees, I see it as ready to merge.

**Comment by jvollhueter** [comment]
@jvollhueter agrees!

**Pull request merged**